### DISCUSSION TITLE
Better support for Type vs. Generic interchange

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#5741] [Language Ideas] [answered]
Currently there are situation where you need to work with a `Type` and on a certain point there is then the requirement to make a `<T>` out of this type. For example if you have a `void *` and the associated `Type` my current way to do this is this (see below).

The problem with this is that the code does not "maintain" itself. I have plenty of code that has the same mechanisms. Introducing the `Half` type was one of the latest large refactorings. I'm havgin different approaches to deal with this:

1. MakeGenericMethod or MakeGericType (-> slow)
2. Creating expression tree and store them in a buffer (-> error prone)

Both of these methods have there own drawbacks so it would be nice to have something in the language that can handle this:
Preferrable would be the possibility to apply a Type instead of <T> as the generic type so that I could write

`Convert<int>` 

or

```cs
Type t = typeof(int)
Convert<t>

```

but I understand that might cause problems. So any other suggestion would be appreciated. A side note: Its in a performance sensitive scenario so I already checked other methods but couldn't find something that helps.

   ```cs
/// <summary>
    /// Converts a memory into a given datatype.
    /// </summary>
    /// <typeparam name="T"></typeparam>
    /// <param name="dataType"></param>
    /// <param name="handle"></param>
    /// <param name="size"></param>
    /// <returns></returns>
    /// <exception cref="NotSupportedException"></exception>
    [SkipLocalsInit, MethodImpl(MethodImplOptions.AggressiveOptimization)]
    private static unsafe void Convert<T>(Type dataType, void* handle, int size, Span<T> targetData) where T : struct
    {
        if (typeof(byte) == dataType)
        {
            var originalData = new ReadOnlySpan<byte>(handle, size);
            ArrayConvert<byte, T>.Cast(originalData, targetData);
        }
        else if (typeof(sbyte) == dataType)
        {
            var originalData = new ReadOnlySpan<sbyte>(handle, size);
             ArrayConvert<sbyte, T>.Cast(originalData, targetData);
        }
        else if (typeof(ushort) == dataType)
        {
            var originalData = new ReadOnlySpan<ushort>(handle, size);
             ArrayConvert<ushort, T>.Cast(originalData, targetData);
        }
        else if (typeof(short) == dataType)
        {
            var originalData = new ReadOnlySpan<short>(handle, size);
             ArrayConvert<short, T>.Cast(originalData, targetData);
        }
        else if (typeof(uint) == dataType)
        {
            var originalData = new ReadOnlySpan<uint>(handle, size);
             ArrayConvert<uint, T>.Cast(originalData, targetData);
        }
        else if (typeof(int) == dataType)
        {
            var originalData = new ReadOnlySpan<int>(handle, size);
             ArrayConvert<int, T>.Cast(originalData, targetData);
        }
        else if (typeof(ulong) == dataType)
        {
            var originalData = new ReadOnlySpan<ulong>(handle, size);
             ArrayConvert<ulong, T>.Cast(originalData, targetData);
        }
        else if (typeof(long) == dataType)
        {
            var originalData = new ReadOnlySpan<long>(handle, size);
             ArrayConvert<long, T>.Cast(originalData, targetData);
        }
        else if (typeof(Half) == dataType)
        {
            var originalData = new ReadOnlySpan<Half>(handle, size);
             ArrayConvert<Half, T>.Cast(originalData, targetData);
        }
        else if (typeof(float) == dataType)
        {
            var originalData = new ReadOnlySpan<float>(handle, size);
             ArrayConvert<float, T>.Cast(originalData, targetData);
        }
        else if (typeof(double) == dataType)
        {
            var originalData = new ReadOnlySpan<double>(handle, size);
             ArrayConvert<double, T>.Cast(originalData, targetData);
        }
        else
        {
            throw new NotSupportedException();
        }
    }
```

**Comment by HaloFour** [comment]
At this point you are already in the land of reflection and the language really doesn't have better options than you do in order to treat the variable of `Type` as a generic type parameter.

That said, for your use case I'm pretty sure there's a built-in way to treat a `void*` as a `Span<T>` of some arbitrary blittable type, but I don't know what it is off of the top of my head.

**Reply by msedi** [reply]
@HaloFour: To create a `Span<T>` from a `void *` is not the big problem as you see in the code above. The problem is that I don't have a `T` ;-)

**Comment by fa1ntl1ght** [comment, marked as answer]
#5267

**Reply by msedi** [reply]
Thx @0x0737: I upvoted your proposal. Lets see whats happening. I fear its too much a niche that some is really interested.

**Reply by fa1ntl1ght** [reply]
I also often come across scenarios where this would be useful (and I guess many others too) and reflection is really too slow for some code, so I hope this will be addressed in the future :) But currently there are many features like DU's and roles & extensions etc. that are obviously higher priority than this, so apparently it will take some releases before something gets done in this space
